compiler/rustc_hir_typeck/src/fallback.rs

```diff
@@ -613,17 +613,21 @@ impl<'tcx> AnnotateUnitFallbackVisitor<'_, 'tcx> {
         if arg_segment.args.is_none()
             && let Some(all_args) = self.fcx.typeck_results.borrow().node_args_opt(id)
             && let generics = self.fcx.tcx.generics_of(def_id)
-            && let args = &all_args[generics.parent_count..]
+            && let args = all_args[generics.parent_count..].iter().zip(&generics.own_params)
             // We can't turbofish consts :(
-            && args.iter().all(|arg| matches!(arg.unpack(), ty::GenericArgKind::Type(_) | ty::GenericArgKind::Lifetime(_)))
+            && args.clone().all(|(_, param)| matches!(param.kind, ty::GenericParamDefKind::Type { .. } | ty::GenericParamDefKind::Lifetime))
         {
+            // We filter out APITs, which are not turbofished.
             let n_tys = args
-                .iter()
-                .filter(|arg| matches!(arg.unpack(), ty::GenericArgKind::Type(_)))
+                .clone()
+                .filter(|(_, param)| {
+                    matches!(param.kind, ty::GenericParamDefKind::Type { synthetic: false, .. })
+                })
                 .count();
-            for (idx, arg) in args
-                .iter()
-                .filter(|arg| matches!(arg.unpack(), ty::GenericArgKind::Type(_)))
+            for (idx, (arg, _)) in args
+                .filter(|(_, param)| {
+                    matches!(param.kind, ty::GenericParamDefKind::Type { synthetic: false, .. })
+                })
                 .enumerate()
```

Comment:
Can you create an iterator and then clone it instead? i.e. `let args = args.filter()` and then `args.clone().count()` + `for ... in args.enumarate()`.

Reply:
good point, can deduplicate the filtering